src/Sparrow.Server/Platform/Posix/CGroupHelper.cs

```diff
@@ -0,0 +1,315 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text.RegularExpressions;
+using System.Threading;
+using Sparrow.Logging;
+using static Sparrow.Server.Platform.Posix.Syscall;
+
+namespace Sparrow.Server.Platform.Posix;
+
+public abstract class CGroup
+{
+    protected static readonly Logger Logger = LoggingSource.Instance.GetLogger("Server", nameof(CGroup));
+    
+    protected const string PROC_CGROUP_FILENAME = "/proc/self/cgroup";
+    protected const string PROC_MOUNTINFO_FILENAME = "/proc/self/mountinfo";
+
+    protected abstract string MemoryLimitFileName { get; } 
+    protected abstract string MemoryUsageFileName { get; } 
+    protected abstract string MaxMemoryUsageFileName { get; }
+
+    private Lazy<CachedPath> _groupPathForMemory;
+    
+    private class CachedPath
+    {
+        public DateTime Time { get; private set; }
+        public string Path { get;}
+
+        public CachedPath(string path, DateTime time)
+        {
+            Path = path;
+            Time = time;
+        }
+
+        public void Deprecate()
+        {
+            Time = DateTime.MinValue;
+        }
+    }
+
+    protected CGroup()
+    {
+        _groupPathForMemory = CreateNewLazyCachedPath();
+    }
+    
+    public long? GetMaxMemoryUsage() => ReadValue(GetGroupPathForMemory, MaxMemoryUsageFileName);
+    public long? GetPhysicalMemoryUsage() => ReadValue(GetGroupPathForMemory, MemoryUsageFileName);
+    public long? GetPhysicalMemoryLimit() => ReadValue(GetGroupPathForMemory, MemoryLimitFileName, CheckLimitValues);
+    
+    private Lazy<CachedPath> GetGroupPathForMemory()
+    {
+        var groupPathForMemory = _groupPathForMemory;
+        if (DateTime.UtcNow - groupPathForMemory.Value.Time > TimeSpan.FromMinutes(1))
+        {
+            var newVal = CreateNewLazyCachedPath();
+            Interlocked.CompareExchange(ref _groupPathForMemory, newVal, groupPathForMemory);
+        }
+
+        return _groupPathForMemory;
+    }
+
+    private Lazy<CachedPath> CreateNewLazyCachedPath()
+    {
+        return new Lazy<CachedPath>(() =>
+        {
+            string path = null;
+            try
+            {
+                path = FindCGroupPathForMemory();
+            }
+            catch (Exception e)
+            {
+                if (Logger.IsOperationsEnabled)
+                    Logger.Operations("Failed to get CGroup path for memory", e);
+            }
+
+            return new CachedPath(path, DateTime.UtcNow);
```

Comment:
Is there is an error, that will return `null`, is the rest of the code ready to handle that?

Reply:
Yes, the rest of the code is checking that
https://github.com/ravendb/ravendb/pull/16339/files#diff-f9e036622d8a33cd68e047f4902e971b72edb349b56cbbe0e42d4395fb43a18eR99